extensions/ticktick/src/utils/date.ts

```diff
@@ -33,7 +33,13 @@ export const formatToServerDate = (date: Date | moment.Moment | null | undefined
   return undefined;
 };
 
-export const formatPrettyDateTime = (date: string) => {
+export const formatPrettyDateTime = (date: string, timezone: string, isFloating: boolean, isAllDay: boolean) => {
   if (!date) return "";
-  return moment(date).format("MMM Do, h:mm a");
+  else if (isAllDay) {
+    return moment(date).format("MMM Do");
+  } else if (isFloating) {
+    return moment(date).format("MMM Do, h:mm a");
+  } else {
+    return moment.tz(date, timezone).format("MMM Do, h:mm a");
+  }
```

Comment:
The `date` property within the task object represents the date and time in UTC format, while the corresponding `timezone` field reflects the timezone of the individual who created the task.

When dealing with all-day events or floating dates — dates without specific times — it is crucial to account for both the original timezone (as determined by the task creator) and the local timezone of the user viewing the task. Conversely, dates that include specific times should be formatted according to the local timezone of the current user.

Reply:
Feel free to conduct your own testing:

- Create an all-day task set in a timezone different from your own. It should appear as occurring on the same day, regardless of your local timezone.
- Create a task with a floating date time, which lacks a specific timezone association. This should be displayed at the same time, irrespective of your local timezone.
- Create a task with a specific time set in a timezone other than your own. It should be displayed with the appropriate time offset relative to your local timezone.

Reply:
Thank you for the review @viduycheung. I've made some changes and think that should cover it. I did try it out but I don't use timezones or floating time setting in normal use so I might have missed something. If I did miss something, any advice on implementation will be really appreciated.

Reply:
You can take a look at this code
``` typescript
export const convertToLocalTimeZoneDate = (
    dateString: string | undefined,
    {
        isAllDay,
        isFloating,
        timeZone
    }: {
        isAllDay?: boolean;
        isFloating?: boolean;
        timeZone?: string;
    }
) => {
    if (!timeZone) {
        return dateString;
    }

    return convertTimeZoneDate(dateString, {
        isAllDay,
        isFloating,
        fromTimeZone: timeZone
    });
};
  
export const convertTimeZoneDate(
    dateString: string | undefined | null,
    {
        isAllDay,
        isFloating,
        fromTimeZone = localTimeZone,
        toTimeZone = localTimeZone
    }: {
        isAllDay?: boolean;
        isFloating?: boolean;
        fromTimeZone?: string;
        toTimeZone?: string;
    }
) => {
    if (!dateString) {
        return null;
    }
    const date = DateUtils.parseUTCString(dateString);
    if (!date) {
        return null;
    }

    if (isAllDay || isFloating) {
        return DateUtils.formatToServerDate(
            convertToLocalDate(date, fromTimeZone, toTimeZone, !isAllDay)
        );
    }

    return dateString;
}

export const convertToLocalDate = (date, fromTimeZone, toTimeZone, hasTime) => {
    const fromMoment = moment.tz(date, fromTimeZone || localTimeZone);
    if (hasTime) {
        return moment
            .tz(
                [
                    fromMoment.year(),
                    fromMoment.month(),
                    fromMoment.date(),
                    fromMoment.hour(),
                    fromMoment.minute()
                ],
                toTimeZone || localTimeZone
            )
            .toDate();
    } else {
        return moment
            .tz(
                [fromMoment.year(), fromMoment.month(), fromMoment.date()],
                toTimeZone || localTimeZone
            )
            .toDate();
    }
};

const startDate = convertToLocalTimeZoneDate(task.startDate);
// format date with converted startDate with the local timezone
moment(startDate).format("xxx")

```

Reply:
Thank you for sharing some code but did you see my new changes? Are they working as expected for you? I just did an update. So if my understanding is correct:
- The datetime is always saved as UTC so we parse the string as UTC
- Then if its an all day task, we convert it to original timezone and then display the day only
- Or if its a floating task, we display the datetime without any conversion
- Lastly if its not floating or all day, we convert to the local timezone and display the datetime

The code you've shown is a bit different and more complex than my understanding so maybe there is some other nuance I have not understood?